diff --git a/psutil/tests/test_linux.py b/psutil/tests/test_linux.py
index 1096d3c..a9cc7ba 100755
--- a/psutil/tests/test_linux.py
+++ b/psutil/tests/test_linux.py
@@ -14,6 +14,7 @@ import errno
 import glob
 import io
 import os
+import platform
 import re
 import shutil
 import socket
@@ -884,6 +885,8 @@ class TestSystemCPUFrequency(PsutilTestCase):
                     self.assertEqual(freq.max, 700.0)
 
     @unittest.skipIf(not HAS_CPU_FREQ, "not supported")
+    @unittest.skipIf(platform.machine() == "aarch64",
+                     "https://github.com/giampaolo/psutil/issues/2373")
     def test_emulate_multi_cpu(self):
         def open_mock(name, *args, **kwargs):
             n = name
