--- ansible-core-2.11.8/examples/ansible.cfg.deprecate	2022-02-06 10:28:35.450116458 -0500
+++ ansible-core-2.11.8/examples/ansible.cfg	2022-02-06 10:28:50.293617139 -0500
@@ -185,6 +185,7 @@
 # To disable these warnings, set the following value to False:
 #
 #deprecation_warnings = True
+deprecation_warnings = False
 
 # Ansible can optionally warn when usage of the shell and
 # command module appear to be simplified by using a default Ansible module
