--- ansible-core-2.13.1/docs/docsite/Makefile.orig	2022-06-22 14:37:19.649929739 -0400
+++ ansible-core-2.13.1/docs/docsite/Makefile	2022-06-22 14:37:42.781216568 -0400
@@ -80,7 +80,7 @@
 ansible_structure:
 	# We must have python and python-packaging for the version_helper
 	# script so use it for version comparison
-	if python -c "import sys, packaging.version as p; sys.exit(not p.Version('$(MAJOR_VERSION)') >  p.Version('2.10'))" ; then \
+	if $(PYTHON) -c "import sys, packaging.version as p; sys.exit(not p.Version('$(MAJOR_VERSION)') >  p.Version('2.10'))" ; then \
 		echo "Creating symlinks in ansible_structure"; \
 		ln -sf ../rst/ansible_index.rst rst/index.rst; \
 		ln -sf ../dev_guide/ansible_index.rst rst/dev_guide/index.rst; \
